base/linalg/diagonal.jl

```diff
@@ -260,8 +262,10 @@ end
 @inline A_mul_Bc(D::Diagonal, rowvec::RowVector) = D*ctranspose(rowvec)
 
 conj(D::Diagonal) = Diagonal(conj(D.diag))
-transpose(D::Diagonal) = D
-ctranspose(D::Diagonal) = conj(D)
+transpose(D::Diagonal{<:Number}) = D
+transpose(D::Diagonal) = Diagonal(transpose.(D.diag))
+ctranspose(D::Diagonal{<:Number}) = conj(D)
+ctranspose(D::Diagonal) = Diagonal(ctranspose.(D.diag))
 
 diag(D::Diagonal) = D.diag
 trace(D::Diagonal) = sum(D.diag)
```

Comment:
OK, so the recursive `ctranspose` or `adjoint` isn't aimed at trying to model an arbitrary block matrix *per se* (with blocks of possibly differing size), but is designed to (for instance) facilitate a matrix of (compatible) linear operators? 

This make a *lot* more sense to me!